redesign: Message actions menu

sayamsamal opened the pull request:
This PR consists of 2 parts:

1. Convert the existing `navbar-dropdown-menu` theme to a common `popover-menu` theme. This change serves as the base for all the other popover redesigns in #28699.
    * Bring the major css properties to the `tippy-box` element itself,
    since that is the parent element of the popovers. This prevents stuff
    like the box-shadow casting shadow on the tippy arrows. Also provides
    for a more consistent styling.
    
    * Added `tippy.js/dist/border.css` and custom css to replace the
    previously used SVG arrow icon implementation for navbar dropdowns.
    This allows the arrows to render properly across all placements.
    
    * Define a minimum width for the popover hotkeys, to allow for a more
    consistent look across different widths of different characters.

2. The redesign part of message-actions-menu.

Fixes: Part of #28699

[CZO thread](https://chat.zulip.org/#narrow/stream/101-design/topic/UI.20redesign.3A.20message.20actions.20menu.20popover.20.2329107/near/1764419)

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

| Desc | Before | After |
|--------|--------|--------|
| Own editable message | ![image](https://github.com/zulip/zulip/assets/82862779/ccb1d6cb-9e6b-4573-b2ac-0ac45125df54) | ![image](https://github.com/zulip/zulip/assets/82862779/2ded6e10-087d-4f35-897a-df3cb35fc6e2) |
| Message by another user | ![image](https://github.com/zulip/zulip/assets/82862779/b7e6cf0e-ba40-4516-a8f5-23ea02ee8bf3) | ![image](https://github.com/zulip/zulip/assets/82862779/8f0b248c-8178-4a4a-a3f6-f7b2db56f65d) |
| Hide muted message again | ![image](https://github.com/zulip/zulip/assets/82862779/5c87ba90-4f39-419d-a0e0-a290e8e46b1a) | ![image](https://github.com/zulip/zulip/assets/82862779/a1824c25-032f-49af-a060-5f872942d69b) |
| In Russian | ![image](https://github.com/zulip/zulip/assets/82862779/57e92688-cf59-4b68-b8af-08a940fb4222) | ![image](https://github.com/zulip/zulip/assets/82862779/d62fd3e2-9543-4d7b-ad6d-650cf5d91460) |
| Dark Mode | ![image](https://github.com/zulip/zulip/assets/82862779/ca8ac1ef-916d-4183-911e-21f040b80b58) | ![image](https://github.com/zulip/zulip/assets/82862779/3b0a25e9-9f02-4e7b-bfe4-badc893fc034) |
| Tooltip for Shift Hotkey Symbol | N/A | ![image](https://github.com/zulip/zulip/assets/82862779/4d6aa531-b4e4-4153-917d-498e9441dd7c) |
| Remove the "View message source icon" | ![image](https://github.com/zulip/zulip/assets/82862779/09672ef0-fd3b-44f8-937a-de61d59deb4d) | ![self-msg-not-edit-or-move](https://github.com/zulip/zulip/assets/82862779/43c29cd7-dc59-40af-85a2-bcd10c05d739) |
| Rename "View message source" in menu options |![image](https://github.com/zulip/zulip/assets/82862779/3c8d3af8-8964-489c-af8c-d2f0d3ddb7d7) | ![view-original-message](https://github.com/zulip/zulip/assets/82862779/a4503d50-5dd0-41af-9fde-60c8f09dc1f3) |

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on web/templates/popovers/actions_popover.hbs:
```diff
@@ -7,7 +7,7 @@
                     <a data-message-id="{{message_id}}" class="respond_button popover-menu-link" tabindex="0">
                         <i class="popover-menu-icon zulip-icon zulip-icon-quote-and-reply" aria-hidden="true"></i>
-                        {{t "Quote and reply" }}
+                        <span class="popover-menu-text">{{t "Quote and reply" }}</span>
```
What's the motivation for adding these spans? I don't see CSS for `popover-menu-text` in this commit; is it just to create a container element?

sayamsamal replied:
Following the new changes, `.popover-menu-text` is now being explicitly used to style the option text.

timabbott replied:
`popover-menu-label` might be a better name for this.

sayamsamal replied:
Renamed to `popover-menu-label`, thanks!

timabbott commented:
I gave this a read and the technical strategy looks broadly reasonable. There's still some design things to settle, including potential changes to icons, though, but I expect it shouldn't feel hard to test-deploy and then merge once we get to a design we're happy with.

timabbott commented on web/src/bundles/app.ts:
```diff
@@ -14,6 +14,7 @@ import "../templates";
 import "../zulip_test";
 
 // Import styles
+import "tippy.js/dist/border.css";
```
Should we have a comment here documenting what these modules do, potentially just in forms to the right upstream links. https://atomiks.github.io/tippyjs/v5/themes/ does not make it obvious what this particular file does.

sayamsamal replied:
I found it in the `v6` docs: https://atomiks.github.io/tippyjs/v6/themes/#arrow-border

sayamsamal replied:
Should we still consider adding a comment linking to the above URL in the tippyjs doc?

sayamsamal replied:
@timabbott a gentle ping towards this.

timabbott replied:
I think it'd be helpful if you split this first commit into two commits -- a first one that adds this and refactors the actual CSS, and a second one that just renames things.

It's really really hard to review commits that do both at the same time, and I think the actual CSS changes are small enough that it should be easy to extract.

sayamsamal replied:
I understand. I have the split the first commit into two smaller commits as suggested. One focusing on the renames, the other on the tippy arrows for the popovers.

sayamsamal replied:
I have also added the reference to https://atomiks.github.io/tippyjs/v6/themes/#arrow-border, along with a small accompanying comment.

sayamsamal commented:
The following changes have been pushed:
1. Add Shift hotkey symbol (⇧) and it's corresponding tooltip.
2. Update Shift hotkey symbol color.
3. Simplify the popover padding, to have a uniform top and bottom padding of `5px` across all popovers.
4. Update from `width: min-content` + `white-space: nowrap` to `width: max-content` + bringing width related properties to the `ul.popover-menu-outer-list` element to support one-liner menu items + wrapping while `max-width` is set.

sayamsamal commented:
Update:
1. Updated the `max-width` logic for popovers, it's now responsive for smaller screen sizes.
2. Fixed the overflow issue caused by setting the `max-width` on the simplebar element instead of an inner element.
  Reference: https://github.com/Grsmto/simplebar/tree/master/packages/simplebar#troubleshooting
> If you are experiencing issues when setting up SimpleBar, it is most likely because your styles are clashing with SimpleBar ones. Make sure the element you are setting SimpleBar on does not override any SimpleBar css properties! We recommend to not style that element at all and use an inner element instead.

sayamsamal commented:
@timabbott this ready for a review.

karlstolley commented on web/styles/app_variables.css:
```diff
@@ -175,12 +175,14 @@
-    --color-background-dropdown-menu: hsl(0deg 0% 100%);
-    --color-border-popover-menu: hsl(0deg 0% 0% / 10%);
+    --color-background-popover-menu: hsl(0deg 0% 100%);
+    --color-border-popover-menu-seperator: hsl(0deg 0% 0% / 10%);
```
_separator_ is so spelled, with an _a_

sayamsamal replied:
Sorry! Fixed the typo and added spell check to my workflow to prevent such inconveniences in the future.

karlstolley commented on web/styles/popovers.css:
```diff
@@ -81,32 +81,39 @@
-            color: var(--color-background-dropdown-menu);
-        }
+    &[data-placement^="top"] > .tippy-arrow::before {
+        bottom: -6.5px;
```
Should these offsets be handled as a variable?

karlstolley commented on web/styles/popovers.css:
```diff
@@ -1224,41 +1202,43 @@ ul {
+        /* Set the popover menu width equal to the width of the
+        longest menu item, thus letting the menu items control
+        the width of the menu. */
         width: min-content;
```
I would think this would be `max-content` here, not `min-content`? But I think it's also possible that the `min-width` and `max-width` declarations here negate the `width` value anyway?

karlstolley replied:
All right :) I think I understand--thanks for the visualization!

sayamsamal replied:
I'm glad it helped, thanks for taking your time to read through it! :)

karlstolley commented on web/styles/popovers.css:
```diff
@@ -1469,17 +1440,26 @@ ul.navbar-dropdown-menu-outer-list {
+
+    > .popover-menu-hotkey-hint {
+        box-sizing: border-box;
+        display: inline-block;
```
Why is it necessary to set `inline-block` here? As a child of `.popover-menu-hotkey-hints`, any `.popover-menu-hotkey-hint` should display as a flex item, rendering its `display` value moot.

sayamsamal replied:
I had put it there to hint towards the fact that the `width` related do apply to the `flex-child` elements. The context for this was that I came across 634a771a37ad475068c1becd29b3dc5c4c2490a4, where this nuance was overlooked (due to which the tooltip hotkeys are no longer squared to a minimum).

karlstolley replied:
Got it. Yes, once something is a flex item, the `width` or `flex-basis` values apply.

karlstolley commented on web/styles/popovers.css:
```diff
@@ -1492,3 +1472,7 @@ ul.navbar-dropdown-menu-outer-list {
 }
+
+#message-actions-menu-dropdown {
+    --popover-menu-max-width: 350px;
```
Why set this variable inside this ID? Would it make better sense to offer this dimension to other popover menus? And keep it with `app_variables.css`?

sayamsamal replied:
To explain the logic behind this, I would need to explain the code snippet from my commit below:

```css
    .simplebar-content {
        /* Set max-width to the value defined by --popover-menu-max-width
        custom property, else fallback to 97vw. */
        max-width: min(var(--popover-menu-max-width, 97vw), 97vw);
    }
```

Here, the intended behavior is this: "We only want a max-width on popovers where we explicitly define it, else we want it to be responsive with the `97vw` value".

In the case that the `--popover-menu-max-width` custom property is not defined (default case), the fallback value of `97vw` is considered, and we get a reduced form `max-width: min(97vw, 97vw); === max-width: 97vw;`.

When we then set the `--popover-menu-max-width` custom property for a certain popover (via targeting its `ID` in this case), we get the reduced form of `max-width: min(--popover-menu-max-width, 97vw);` where `--popover-menu-max-width` comes into play at the desktop sizes, and the `97vw` value comes into play at mobile sizes.

sayamsamal replied:
Please let me know, if you would prefer an alternative to this behavior, thanks!

timabbott replied:
This approach seems OK, but let's make sure to document the motivation for the localized variable in a comment.

sayamsamal replied:
```css
/* Set `max-width` to the minimum of `97vw` and the value defined by
the `--popover-menu-max-width` custom property, which should be defined
as a local custom property in the scope of a particular popover where
we want apply a `max-width` limit.

When we don't define this local custom property, the fallback value of
`97vw` is considered and the `max-width` is then set to `min(97vw, 97vw)`
which is simply equivalent to `97vw`.

IMPORTANT: Setting `--popover-menu-max-width` in the `:root` scope
instead of the local scope, eliminates the fallback behavior and thus
prevents us from setting the max-width on a case-to-case basis. */
max-width: min(var(--popover-menu-max-width, 97vw), 97vw);
```

This is the comment I came up with, to document the motivation for the localized variable.

sayamsamal commented:
@karlstolley Requesting another review from you. I have incorporated your feedback, and attempted to explain my logic with respect to the popover width with some more clarity.

timabbott commented on web/styles/popovers.css:
```diff
@@ -81,32 +81,39 @@
+
+    &[data-placement^="right"] > .tippy-arrow::before {
+        left: -6.5px;
+    }
```
We should also have comments explaining what these are doing, since it's overriding a bit of upstream CSS, I imagine.

timabbott replied:
Ahh, OK, I see you add this in a later commit.

sayamsamal replied:
Yes! Would you like me to merge the later commit with the former one?

timabbott replied:
Nah, this is fine, it's just documentation.

timabbott commented on web/src/common.ts:
```diff
@@ -95,7 +95,7 @@ export function adjust_mac_kbd_tags(kbd_elem_class: string): void {
 // We convert the hotkey hints used in the tooltips to mac equivalent
 // key combinations, when we detect that the user is using a mac-style keyboard.
-export function adjust_mac_tooltip_keys(hotkeys: string[]): void {
+export function adjust_mac_hotkeys(hotkeys: string[]): void {
```
Would `adjust_mac_hotkey_hints` be a better name?

sayamsamal replied:
Sounds good to me; renamed to `adjust_mac_hotkey_hints`, thanks!

timabbott commented on web/styles/app_variables.css:
```diff
@@ -495,7 +498,6 @@
         hsl(0deg 0% 11%)
     );
     --color-background-unread-counter-dot: hsl(105deg 2% 50% / 90%);
-    --color-background-unread-counter-popover-menu: hsl(105deg 2% 50% / 50%);
```
These removals should be in the commit that removes that icon use, I think?

sayamsamal replied:
Fixed. Sorry for the mistake, I think I made the error while squashing commits.

timabbott commented:
Posted a handful of comments, otherwise lgtm.

sayamsamal commented:
I have addressed the feedback, @timabbott could you take a look? Also I'd like to grab your attention to: https://github.com/zulip/zulip/pull/29107#discussion_r1551914857.

timabbott commented:
Changes look great, but there's a few comment threads still needing a response from you.

sayamsamal commented:
@timabbott I have now also addressed the missed feedback. Requesting another review from you.

timabbott commented:
Merged, huge thanks for all the work on this @sayamsamal!
